File: 0540-single-element-in-a-sorted-array/0540-single-element-in-a-sorted-array.java
class Solution {
    public int singleNonDuplicate(int[] nums) {
        int n = nums.length;
        
        // Case 1
        if(n==1){
            return nums[0];
        }
        
        // Case 2
        if(nums[0] != nums[1]){
            return nums[0];
        }
        
        // Case 3 
        if(nums[n-1] != nums[n-2]){
            return nums[n-1];
        }
        
        // Case 4
        int start = 1;
        int end = n-2;
        
        while(start<=end){
            int mid = start + (end-start)/2;
            
            if(nums[mid] != nums[mid-1] && nums[mid]!=nums[mid+1]){
                return nums[mid];
            }
            
            // Eleminating left half
            if((mid%2 == 1 && nums[mid - 1] == nums[mid]) || mid%2==0 && nums[mid] == nums[mid+1]){
                start = mid+1;
            }else{
                // Eleminating right half
                end = mid-1;
            }
        }
        return -1;
    }
}

File: 0540-single-element-in-a-sorted-array/0540-single-element-in-a-sorted-array.cpp
class Solution {
public:
    int found(vector<int> arr, int n){
        // case 1
        if(n==1) return arr[0];
        
        // case 2
        if(arr[0] != arr[1]) return arr[0];
        
        // case 3
        if(arr[n-1]!=arr[n-2]) return arr[n-1];
        
        // case 4
        int low = 1;
        int high = n-1;
        while(low<=high){
            int mid = (low+high)/2;
            if(arr[mid] != arr[mid-1] && arr[mid] != arr[mid+1]){
                return arr[mid];
            }
            
            if((mid%2==1 && arr[mid] == arr[mid-1]) || (mid%2==0 && arr[mid] == arr[mid+1])){
                low = mid+1;
            }else{
                high = mid-1;
            }
        }
        
        return -1;
    }
    int singleNonDuplicate(vector<int>& nums) {
        return found(nums, nums.size());
    }
};